### PR TITLE
task 4 mentor cheack

### PR DESCRIPTION
**Pull request by BondPV**
[deploy](https://bondpv-rss-react.netlify.app)

[task](https://github.com/rolling-scopes-school/tasks/tree/master/react/modules/module04)

READY:
- Add API https://rickandmortyapi.com/documentation/
- All the functionality should work.
- Add tests. All the api-calls should be mocked in the tests.
- Test coverage for the whole app should be not less than 70%.

### CODE REVIEW

File: src/components/Preloader/Preloader.module.scss [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  60% {
+    right: -90%;
+    left: 100%;
+  }
+
+  100% {
+    right: -90%;
+    left: 100%;
+  }
+}
```
Comment (Ksarelto) [inline src/components/Preloader/Preloader.module.scss:56]: check empty spaces) and why u delete stylelint?

File: src/components/CardModal/CardModal.module.scss [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  margin: 1.5rem 3rem 0;
+}
+
+.card {
+  position: relative;
+  width: 400px;
+  animation: show 0.6s ease-in-out;
+  background-color: #fff;
+
+  @include box-shadow;
```
Comment (Ksarelto) [inline src/components/CardModal/CardModal.module.scss:30]: if u include this to '.card' then include should be first in styles list

File: src/components/Card/Card.test.tsx [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { screen, render } from '@testing-library/react';
 import { Card } from './Card';
 import { ICard } from '../../types/interfaces';
-import database from '../../database/source.json';
+import database from '../../Api/mockSource.json';
```
Comment (Ksarelto) [inline src/components/Card/Card.test.tsx:4]: better create separate folder 'mocks' where u will collect all your mock objects and functions

File: src/components/Card/Card.tsx [hunk 4]
```diff
@@ -1,27 +1,23 @@
...
       </div>
-      <div className={styles.card__episode}>
-        <span className={styles.card__count}>{episode.length}</span>
-        <em>episode count</em>
-      </div>
-    </div>
-  </div>
-);
-
+      {isCardModalShow ? <CardModal id={id} handleCardModalShow={handleCardModalShow} /> : ''}
```
Comment (Ksarelto) [inline src/components/Card/Card.tsx:19]: u can use just like: 
```
{isCardModalShow && <CardModal id={id} handleCardModalShow={handleCardModalShow} />}
```
false will not be rendered

File: src/components/Alert/Alert.test.tsx [hunk 5]
```diff
@@ -1,9 +1,11 @@
 import { render, screen } from '@testing-library/react';
+import { vi } from 'vitest';
 import { Alert } from './Alert';
 
 describe('Alert message', () => {
-  test('should render Alert component', () => {
-    render(<Alert message="Alert message" isShow={true} setIsShow={() => {}} />);
+  test('should render', () => {
```
Comment (Ksarelto) [inline src/components/Alert/Alert.test.tsx:6]: please, add what this test should render

File: src/Api/Api.test.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
+import { rest } from 'msw';
+import { setupServer } from 'msw/node';
+import { getCharacters, getCharacter } from './Api';
+import mockResponse from './mockSource.json';
+
+describe('Api function getCharacters', () => {
+  const params = { page: 1, name: '' };
+
+  const server = setupServer(
```
Comment (Ksarelto) [inline src/Api/Api.test.tsx:9]: do u need this according to your task?

File: src/Api/Api.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  next: string | null;
+  prev: string | null;
+}
+
+interface IResponseCharacters {
+  info: IResponseInfo;
+  results: ICard[];
+}
+
+const API_URL = 'https://rickandmortyapi.com/api/character';
```
Comment (Ksarelto) [inline src/Api/Api.tsx:15]: usually this url is stored in env but for this small api u can save it here

File: src/Api/Api.tsx [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+const getCharacters = async (params: ISearchParams): Promise<ICard[] | null> => {
+  const queryString = Object.entries(params)
```
Comment (Ksarelto) [inline src/Api/Api.tsx:18]: u can use here:
```
const queryString = new URLSearchParams(params as unknown as URLSearchParams).toString()
```

File: src/components/CardModal/CardModal.test.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import { vi } from 'vitest';
+import { CardModal } from './CardModal';
+import { ICard } from '../../types/interfaces';
+import database from '../../Api/mockSource.json';
+
+describe('CardModal', () => {
+  const mockCard: ICard = database.results[0];
+
+  test('should render the card modal', () => {
+    const handleClose = vi.fn();
```
Comment (Ksarelto) [inline src/components/CardModal/CardModal.test.tsx:11]: if u use the same function in several places move her out of the describe and reuse in other places

File: src/components/CardModal/CardModal.test.tsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+    render(<CardModal id={mockCard.id} handleCardModalShow={handleClose} />);
```
Comment (Ksarelto) [inline src/components/CardModal/CardModal.test.tsx:13]: the same u can try to do with render

File: src/components/CardModal/CardModal.tsx [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+    if (cardInfo) {
+      setLoading(false);
+      setCard(cardInfo);
+    }
+  };
+
+  const [createdDate] = card.created?.split('T') || '';
+
+  const notClick = (e: React.MouseEvent): void => e.stopPropagation();
```
Comment (Ksarelto) [inline src/components/CardModal/CardModal.tsx:37]: what do u mean 'notClick' ?))

File: src/components/Preloader/Preloader.module.scss (hunk 1 above)
Comment (BondPV) [reply src/components/Preloader/Preloader.module.scss:56]: add spaces

File: src/components/CardModal/CardModal.module.scss (hunk 2 above)
Comment (BondPV) [reply src/components/CardModal/CardModal.module.scss:30]: fixed

File: src/components/Card/Card.tsx (hunk 4 above)
Comment (BondPV) [reply src/components/Card/Card.tsx:19]: fixed

File: src/Api/Api.tsx (hunk 8 above)
Comment (BondPV) [reply src/Api/Api.tsx:18]: thanks, didn't know

File: src/components/CardModal/CardModal.test.tsx (hunk 9 above)
Comment (BondPV) [reply src/components/CardModal/CardModal.test.tsx:11]: fixed

File: src/components/CardModal/CardModal.test.tsx (hunk 10 above)
Comment (BondPV) [reply src/components/CardModal/CardModal.test.tsx:13]: added beforeEach

File: src/components/CardModal/CardModal.tsx (hunk 11 above)
Comment (BondPV) [reply src/components/CardModal/CardModal.tsx:37]: renamed function

**Pull request merged**